File: go/registry.go
package gobutton

// Package gobutton provides functions for registering and unregistering
// buttons with the button-hub API.

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

// Register tells button-hub to recognise the given button_id.
// It is best-effort: failure is logged but not fatal.
func Register(ctx context.Context, apiURL, buttonID string) error {
	url := strings.TrimSuffix(apiURL, "/") + "/buttons"
	body, _ := json.Marshal(map[string]string{"button_id": buttonID})

	req, err := http.NewRequestWithContext(ctx, "POST", url, bytes.NewReader(body))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		rb, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("register failed: %d %s", resp.StatusCode, string(rb))
	}
	return nil
}

// Unregister removes the button_id from button-hub.
func Unregister(ctx context.Context, apiURL, buttonID string) error {
	url := fmt.Sprintf("%s/buttons/%s", strings.TrimSuffix(apiURL, "/"), buttonID)

	req, err := http.NewRequestWithContext(ctx, "DELETE", url, nil)
	if err != nil {
		return err
	}

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		rb, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("unregister failed: %d %s", resp.StatusCode, string(rb))
	}
	return nil
}

File: crates/client/src/registry.rs
//! Button registration and unregistration with the hub API.

use std::error::Error;

/// Registers a button ID with button-hub.
///
/// This is best-effort — errors are logged but not fatal.
pub async fn register(api_url: &str, button_id: &str) -> Result<(), Box<dyn Error + Send + Sync>> {
    let url = format!("{}/buttons", trim_slash(api_url));
    let body = serde_json::json!({ "button_id": button_id });

    let client = reqwest::Client::new();
    let resp = client
        .post(&url)
        .json(&body)
        .send()
        .await?;

    if !resp.status().is_success() {
        let status = resp.status();
        let body = resp.text().await.unwrap_or_default();
        return Err(format!("register failed: {} {}", status, body).into());
    }

    Ok(())
}

/// Unregisters a button ID from button-hub.
pub async fn unregister(
    api_url: &str,
    button_id: &str,
) -> Result<(), Box<dyn Error + Send + Sync>> {
    let url = format!("{}/buttons/{}", trim_slash(api_url), button_id);

    let client = reqwest::Client::new();
    let resp = client.delete(&url).send().await?;

    if !resp.status().is_success() {
        let status = resp.status();
        let body = resp.text().await.unwrap_or_default();
        return Err(format!("unregister failed: {} {}", status, body).into());
    }

    Ok(())
}

fn trim_slash(url: &str) -> &str {
    url.trim_end_matches('/')
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_trim_slash() {
        assert_eq!(trim_slash("http://localhost:3000/"), "http://localhost:3000");
        assert_eq!(trim_slash("http://localhost:3000"), "http://localhost:3000");
        assert_eq!(trim_slash("http://localhost:3000//"), "http://localhost:3000");
    }
}